pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+			// We don't care about `done_amount` until swap into foreign is fulfilled
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone { swap, .. } => {
+				InvestmentState::<T>::insert(who, investment_id, state);
+				Ok((state, Some(swap), Zero::zero()))
+			},
+			InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing { swap, invest_amount } |
+			InvestState::ActiveSwapIntoForeignCurrencyAndInvestmentOngoing { swap, invest_amount } |
+			// We don't care about `done_amount` until swap into foreign is fulfilled
+			InvestState::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing { swap,invest_amount, .. } => {
+				InvestmentState::<T>::insert(who, investment_id, state);
+				Ok((state, Some(swap), invest_amount))
+			},
+			InvestState::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDone { swap, done_amount } => {
+				maybe_executed_decrease = Some((swap.currency_out, done_amount));
+
+				let new_state = InvestState::ActiveSwapIntoPoolCurrency { swap };
+				InvestmentState::<T>::insert(who, investment_id, new_state);
+
+				Ok((new_state, Some(swap), Zero::zero()))
+			},
+			InvestState::ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing { swap, done_amount, invest_amount } => {
+				maybe_executed_decrease = Some((swap.currency_out, done_amount));
+
+				let new_state = InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing { swap, invest_amount };
+				InvestmentState::<T>::insert(who, investment_id, new_state);
+
+				Ok((new_state, Some(swap), invest_amount))
+			},
+			InvestState::SwapIntoForeignDone { done_swap } => {
+				maybe_executed_decrease = Some((done_swap.currency_in, done_swap.amount));
+
+				InvestmentState::<T>::remove(who, investment_id);
+
+				Ok((InvestState::NoState, None, Zero::zero()))
+			},
+			InvestState::SwapIntoForeignDoneAndInvestmentOngoing { done_swap, invest_amount } => {
+				maybe_executed_decrease = Some((done_swap.currency_in, done_swap.amount));
+
+				let new_state = InvestState::InvestmentOngoing { invest_amount };
+				InvestmentState::<T>::insert(who, investment_id, new_state);
+
+				Ok((new_state, None, invest_amount))
+			},
+		}
+		.map(|(invest_state, maybe_swap, invest_amount)| {
+			let (maybe_invest_state_prio, maybe_new_redeem_state) = Self::handle_swap_order(who, investment_id, maybe_swap,  TokenSwapReason::Investment)?;
+
+			// Dispatch transition event, post swap state has priority if it exists as it was the last transition
+			if let Some(invest_state_prio) = maybe_invest_state_prio {
+				Self::deposit_investment_event(who, investment_id, Some(invest_state_prio));
+			} else {
+				Self::deposit_investment_event(who, investment_id, Some(invest_state));
+			}
+			Self::deposit_redemption_event(who, investment_id, maybe_new_redeem_state);
+
+			// Update investment after all states have been updated
+			T::Investment::update_investment(who, investment_id, invest_amount)?;
```

Comment:
Maybe we can save some read access and event generation from `pallet-investment` calling this method only here if `invest_amount > 0 `

Reply:
Actually, the only reason to not call `T::Investment::update_investment` would be if the `invest_amount` is unchanged  from the `InvestState` before applying transition(s) and after.

When we update an investment to the zero amount, we reduce the unprocessed investment amount of zero which would be desired in this circumstance.

Reply:
Ahh yes! Totally right! I was checking that a first `increase_foreign_investment()` was calling to it always with zero. But it's true that from your pallet perspective, it's difficult to check.